# Introduction

This is a continuation of the `dice_sample` and `dice_posterior` assignments.
In these assignments, we have a bag containing two types of dice with different
probabilities of rolling each number (also referred to as a "face" of the die).
Someone selects a die from the bag at random, rolls it a fixed number of times,
reports the outcomes, returns it to the bag, and repeats the process. We refer
each selection of a new die as a "draw". Here, you will write code to run the EM
(Expectation Maximization) algorithm to estimate the parameters of the system --
the probability of drawing each die type and the conditional probability of each
face given the die type.

This notebook provides a brief overview of the assignment but you'll have to
read the code for detailed instructions.  

All of the graded tests are written such that only two dice are in the bag.
Feel free write the code such that it only functions on two dice or to write 
more general code which can operate on bags of any composition.

## Die and BagOfDice classes
For this assignment, you will use our implementation of these two classes. These 
lines at the top of the assignment file make those two classes availble from our 
cse587Autils package:
```python
from cse587Autils.DiceObjects.Die import Die
from cse587Autils.DiceObjects.BagOfDice import BagOfDice
```
These are more full featured implementations than the code we provided with
the `dice-posterior` assignment.


### Die class
Instantiate with a list of the probabilities of rolling each face on the die.
The list can be of any length but the probabilities must sum to one.
```python
biased_die = Die(face_probs=[0.1, 0.2, 0.7])
print(biased_die) # prints Die([0.1, 0.2, 0.7])
```
You can get the number of faces with the <len> function
```python
len(biased_die) # returns 3
```
Previously, we used `num_faces` for that. You can also index into the die object
directly to get face probabilities:

```python
biased_die[1] # returns 0.2. Note zero-based indexing.
```

### BagOfDice class
Instantiate with two arguments. The first a list of the probabilies of drawing
each die type, which must sum to one. The second is a list of Die objects 
representing the unique die types in the bag (each type is listed only once).

In [8]:
from cse587Autils.DiceObjects.BagOfDice import BagOfDice, Die
fair_die = Die([1/3]*3) # In python, this is equivalent to Die[1/3, 1/3, 1/3]
biased_die = Die(face_probs=[0.1, 0.2, 0.7])

bag = BagOfDice([0.25, 0.75],[fair_die, biased_die])
print(bag)

BagOfDice(die_priors=[0.25, 0.75], dice=[Die([0.3333, 0.3333, 0.3333]), Die([0.1, 0.2, 0.7])])


Indexing into the bag returns tuple of the corresponding die's probability of
being drawn and the die object itself. You can also iterate through the dice.

In [12]:
bag[1] # Returns (0.75, Die([0.1, 0.2, 0.7]))
for prior_die_pair in bag:
    die_prior, die = prior_die_pair
    print(f'die prior: {die_prior}\ndie object: {die}')

die prior: 0.25
die object: Die([0.3333, 0.3333, 0.3333])
die prior: 0.75
die object: Die([0.1, 0.2, 0.7])


BagOfDice supports the following operations:

In [3]:
# len() returns the number of dice in the bag
print(f'len() returns the number of dice in the bag: {len(bag)}')

# there are getters (and setters) for the die_priors and dice
print(f'accessing the die priors: {bag.die_priors}')
print(f'accessing the Die objects: {bag.dice}')

len() returns the number of dice in the bag: 2
accessing the die priors: [0.25, 0.75]
accessing the Die objects: [Die([0.3333, 0.3333, 0.3333]), Die([0.1, 0.2, 0.7])]


We can use the BagOfDice to generate sample data. For example, to 
produce a sample by drawing a die, with replacement, from the bag 5 times 
and rolling each die drawn 20 times, we can use the draw() method. `draw()`
returns a numpy array containg the number of times each face was rolled on a 
given draw. The array length is the maximum number of faces among all the dice 
in the bag. Faces that were not rolled (either by chance or because they don't 
exist on the die) are represented by 0.

In [70]:
sample_data = [bag.draw(20) for _ in range(5)]
print(sample_data)

[array([3, 8, 9]), array([ 1,  8, 11]), array([ 1,  6, 13]), array([ 1,  4, 15]), array([ 8, 10,  2])]


In your EM implementation, you will use a BagOfDice object to store the current 
estimates of the parameters. On each call to `m_step` you will create a new bag
with updated parameter estimations.

Your EM algorithm must stop when it converges -- that is, when the change in 
parameter estimates from one iteration to the next becomes very small. To
characterize the change in parameters, we will use the sum of the absolute values
of the differences between between the parameters, include die priors and face
probabilities. We have implemented this measure for you. To get it, simply use
the subtraction operator on two BagOfDice objects, as shown below. This will 
compare the first die of one bag to the first die of the other, etc. The order
of dice is assumed to be the same, but this is not a problem for EM iterations --
your estimates for each die will always be in the same order.

In [71]:
bag1 = BagOfDice([0.5, 0.5], 
                 [Die([0.9, 0.1, 0.0]), Die([0.1, 0.1, 0.8])])
bag2 = BagOfDice([0.6, 0.4], 
                 [Die([0.9, 0.1, 0.0]), Die([0.1, 0.2, 0.7])])
print(f'bag difference: {bag1 - bag2}')

bag difference: 0.4


## Input and Output

The top level function is called diceEM:

```python
def diceEM(experiment_data: List[NDArray[np.int_]],
           bag_of_dice: BagOfDice,
           accuracy: float,
           max_iterations: int = int(1e4)) -> [int, BagOfDice]:
```

where,

- `experiment_data`  is a list of draws, each of which gives the results of
drawing a die from the bag and rolling it n times and then aggregating the
number of times each face appears. It is a list of numpy arrays, where the number 
of arrays is equal to the number of `draws` that is performed, and the sum of
each individual array is the total number of times the die was rolled on that
draw.

- `max_iterations` similarly sets a threshold at the number of iterations the EM
algorithm can run -- it is a good idea to set this to avoid endless, or
needlessly long, loops. But, you may need to adjust this value depending on the
accuracy. Default is 10000.

The number of faces, trials per draw, etc., can all be calculated from the input
data and so will not be provided. 

`diceEM()` implements the outer loop of the EM algorithm. It calls `e_step()`
and `m_step()` on each iteration and updates its estimate of the parameters
which generated the `experiment_data`. This updated estimate is used in the
sebsequent iteration, and so on until the stop conditions are met.

The return value of `diceEM()` is the final estimate of the parameters in the 
form of a `BagOfDice` object.

Your `e_step()` code needs to calculate posteriors. You are welcome to use your 
dice_posterior code. If you did *not* have functioning `dice_posterior()` code, 
or you question its accuracy after pasting it in to this assignment, the TAs can
provide correct dice_posterior code to you.

Outlines of the code are provided in the file [assignment.py](assignment.py).
Read the comments, too. You need to fill in key parts of the algorithms. Feel 
free to paste your dice_posterior code in.  

**Important**: your code will be graded against the
[test_assignment.py](test_assignment.py) code exactly as it is when you received
this assignment. You are welcome to change it in your project, but those changes
will *not* be reflected when your code is graded. So, we suggest you *do not*
change the tests. Otherwise, you will potentially be surprised if your code
fails during the automated grading.

## EM Initialization Tips

To initialize the parameter estimates, do not make all the possibilities
equally likely. If you do, the algorithm may get stuck and take longer to
converge. However, do not make them
too far from equally likely, either, to avoid strongly biasing the final result
by the initial values. Since there are only two die types, I suggest
initializing their probabilities to 0.45 and 0.55. For the probabilities of the
faces of each die, I took a random real between 1/n and 2/n, where n is the
number of faces. Then I normalized them so they would all add up to one using the call

```
list_of_numbers / sum(list_of_numbers)
```

## Experiments and questions

### Number of iterations needed for convergence

Given a very large dataset, how many iterations does it take to converge to within the required accuracy of 10^-4? How does that change if you tighten the convergence criterion to 10^-6? How about 10^-8. Please make a comment on the general pattern you observe.

In [77]:
from importlib import reload
import diceEM.assignment as assign
reload(assign) # This ensures code is reloaded each time
from diceEM.assignment import diceEM, generate_sample

experiment_result = generate_sample([4,6], 
                                    [[0.3, 0.2, 0.2, 0.3], 
                                     [0, 0.2, 0.2, 0.6]], 
                                     4000, 
                                     30)

init_bag = BagOfDice([0.75, 0.25], 
                     [Die([0.1, 0.3, 0.5, 0.1]), 
                      Die([0.1, 0.3, 0.4, 0.2])])
diceEM(experiment_result, init_bag, 10e-8, 100)

(12,
 BagOfDice(die_priors=[0.40900946601771687, 0.5909905339822831], dice=[Die([0.2964, 0.1988, 0.202, 0.3028]), Die([0.0, 0.1994, 0.1987, 0.6019])]))

Answer: It takes 6 iterations to converge to within the accuracy of 10^-4 for this large dataset. As for a accuracy of 10^-6 and 10^-8, it takes 9 and 12 iterations to converge, respectively. Thus, the number of iterations to converge is increasing as accuracy is tightened. This is expected because more iterations are required for a more stringent convergence. Also, it seems that the iterations required doesn't dramatically increase upon tightening the accuracy by 2 or 4 orders of magnitude, which means that the algorithm is quite efficient in converging to the local optimum.

How much effect does tightening the accuracy requirement by 2 or 4 orders of magnitude have on how well the algorithm identifies the correct parameters?

Answer:

When accuracy = 10^-4, BagOfDice(die_priors=[0.40896643154383977, 0.5910335684561603], dice=[Die([0.2964, 0.1988, 0.202, 0.3028]), Die([0.0, 0.1994, 0.1987, 0.6018])])

When accuracy = 10^-6, BagOfDice(die_priors=[0.4090090197676536, 0.5909909802323464], dice=[Die([0.2964, 0.1988, 0.202, 0.3028]), Die([0.0, 0.1994, 0.1987, 0.6019])])

When accuracy = 10^-8, BagOfDice(die_priors=[0.40900946601771687, 0.5909905339822831], dice=[Die([0.2964, 0.1988, 0.202, 0.3028]), Die([0.0, 0.1994, 0.1987, 0.6019])])

Thus, tightening the accuracy requirements by 2 or 4 orders of magnitude doesn't have a significant effect on finding the correct parameters because all the resulting parameters shown above are almost the same. Also, all the converged parameters are close to the true value used in sample generation.

Now let's change the parameters of the sample generated as shown below. 

How does this affect the number of iterations required to make the accuracy goal? Why do you think that is?

In [78]:
from importlib import reload
import diceEM.assignment as assign
reload(assign) # This ensures code is reloaded each time
from diceEM.assignment import diceEM, generate_sample

experiment_result = generate_sample([4, 6], [[0.3, 0.2, 0.3, 0.2], 
                                             [0.1, 0.2, 0.2, 0.5]], 
                                             4000, 
                                             30)

init_bag = BagOfDice([0.75, 0.25], [Die([0.1, 0.3, 0.5, 0.1]), 
                                    Die([0.1, 0.3, 0.4, 0.2])])

diceEM(experiment_result, init_bag, 10e-4, 100)

(9,
 BagOfDice(die_priors=[0.40416219790208086, 0.5958378020979191], dice=[Die([0.2973, 0.1988, 0.304, 0.2]), Die([0.1004, 0.1993, 0.1998, 0.5005])]))

Answer:

It takes 9 iterations to converge, which is larger than the previous case (6 iterations). The original distribution for the second die had a clear distinction with a 0 probability for the first outcome (i.e., no chance of rolling a 1). In the changed code, this probability is now 0.1, which could increase the overlap between the two dice's outcome distributions. When the outcome probabilities for the two dice become more similar, it becomes harder for the EM algorithm to distinguish between the dice, requiring more iterations to accurately estimate the parameters.

### Sample size

1. Consider the following experiment.

In [79]:
from importlib import reload
import diceEM.assignment as assign
reload(assign) # This ensures code is reloaded each time
from diceEM.assignment import diceEM, generate_sample

face_probs = [[0.3, 0.3, 0.2, 0.2],
              [0.2, 0.2, 0.2, 0.4]]

experiment_result = generate_sample([4, 6], face_probs, 6000, 30)
init_bag = BagOfDice([0.75, 0.25], [Die([0.1, 0.3, 0.5, 0.1]), 
                                    Die([0.1, 0.3, 0.4, 0.2])])

inference_result = diceEM(experiment_result, init_bag, 10e-4, 100)
print(inference_result[1])

BagOfDice(die_priors=[0.40254223047228116, 0.5974577695277188], dice=[Die([0.2986, 0.3027, 0.2015, 0.1972]), Die([0.1988, 0.1986, 0.2011, 0.4015])])


- How many die rolls are in this sample? 

Answer: 30 per draw and 6000 draws give a total of 180000 rolls.

- Try reducing the number of draws/trials while keeping the number of rolls constant. How many draws do you need with 30 rolls each to get the right answer consistently, to within two decimal places? 

In [98]:
for data_seed in range(10):
    data = generate_sample([4, 6], face_probs, 2500, 30, seed=data_seed)
    print(diceEM(data, init_bag, 10e-4, 100))

(24, BagOfDice(die_priors=[0.3991593686092596, 0.6008406313907404], dice=[Die([0.3037, 0.2999, 0.1965, 0.1999]), Die([0.2017, 0.199, 0.202, 0.3973])]))
(26, BagOfDice(die_priors=[0.39464281093822207, 0.6053571890617778], dice=[Die([0.2979, 0.2972, 0.202, 0.2029]), Die([0.2031, 0.1993, 0.2013, 0.3963])]))
(24, BagOfDice(die_priors=[0.391495855955232, 0.608504144044768], dice=[Die([0.3102, 0.2973, 0.195, 0.1976]), Die([0.2027, 0.201, 0.2007, 0.3955])]))
(24, BagOfDice(die_priors=[0.4052832364622341, 0.5947167635377659], dice=[Die([0.302, 0.2964, 0.1982, 0.2034]), Die([0.1972, 0.2009, 0.2006, 0.4013])]))
(25, BagOfDice(die_priors=[0.39853011605977373, 0.6014698839402263], dice=[Die([0.3017, 0.2953, 0.1979, 0.2051]), Die([0.2015, 0.1998, 0.1974, 0.4013])]))
(22, BagOfDice(die_priors=[0.3929973084138628, 0.6070026915861372], dice=[Die([0.2978, 0.2996, 0.2038, 0.1988]), Die([0.2002, 0.1949, 0.202, 0.4029])]))
(24, BagOfDice(die_priors=[0.4078485551594918, 0.5921514448405082], dice=[Die([0.30

Answer: 2500 draws are needed to give a fairly consistent result.

- If you don't have quite enough draws/trials, which parameters tend to be off?

The priors were more off than the face probs.

- Now try changing the number of rolls per trial to 50. How many trials do you 
need to consistently get the right  answer consistently, to within two decimal places? 

In [100]:
for data_seed in range(10):
    data = generate_sample([4, 6], face_probs, 1500, 50, seed=data_seed)
    print(diceEM(data, init_bag, 10e-4, 100))

(14, BagOfDice(die_priors=[0.41272766282967543, 0.5872723371703246], dice=[Die([0.3058, 0.2967, 0.1973, 0.2002]), Die([0.1997, 0.1997, 0.2027, 0.3979])]))
(14, BagOfDice(die_priors=[0.40383966674186716, 0.5961603332581329], dice=[Die([0.302, 0.2947, 0.2041, 0.1992]), Die([0.201, 0.1963, 0.2008, 0.4019])]))
(14, BagOfDice(die_priors=[0.41361274061189907, 0.586387259388101], dice=[Die([0.3026, 0.2994, 0.1983, 0.1997]), Die([0.2009, 0.202, 0.1963, 0.4007])]))
(14, BagOfDice(die_priors=[0.40463722975130817, 0.5953627702486918], dice=[Die([0.2973, 0.2982, 0.2, 0.2045]), Die([0.1987, 0.201, 0.1972, 0.4032])]))
(14, BagOfDice(die_priors=[0.38340539284070085, 0.6165946071592993], dice=[Die([0.3003, 0.2985, 0.1955, 0.2057]), Die([0.2042, 0.1994, 0.1957, 0.4007])]))
(13, BagOfDice(die_priors=[0.3767662494130569, 0.6232337505869431], dice=[Die([0.2981, 0.3019, 0.2004, 0.1996]), Die([0.2001, 0.1976, 0.2017, 0.4007])]))
(15, BagOfDice(die_priors=[0.415919967444838, 0.5840800325551619], dice=[Die([0

Answer: 1500 draws are needed.

- Does tightening the convergence criterion make the results substantially more accurate? 

In [102]:
for data_seed in range(10):
    data = generate_sample([4, 6], face_probs, 1500, 50, seed=data_seed)
    print(diceEM(data, init_bag, 10e-8, 100))

(30, BagOfDice(die_priors=[0.41310753273777107, 0.5868924672622289], dice=[Die([0.3058, 0.2966, 0.1973, 0.2003]), Die([0.1996, 0.1997, 0.2027, 0.398])]))
(29, BagOfDice(die_priors=[0.4040641255591735, 0.5959358744408265], dice=[Die([0.302, 0.2947, 0.2041, 0.1992]), Die([0.201, 0.1963, 0.2008, 0.402])]))
(29, BagOfDice(die_priors=[0.4138819862429429, 0.5861180137570571], dice=[Die([0.3026, 0.2993, 0.1983, 0.1997]), Die([0.2009, 0.202, 0.1963, 0.4007])]))
(31, BagOfDice(die_priors=[0.4050538347108179, 0.594946165289182], dice=[Die([0.2973, 0.2982, 0.2, 0.2046]), Die([0.1986, 0.2009, 0.1972, 0.4032])]))
(31, BagOfDice(die_priors=[0.38381966901293785, 0.6161803309870622], dice=[Die([0.3002, 0.2985, 0.1955, 0.2058]), Die([0.2041, 0.1994, 0.1957, 0.4008])]))
(27, BagOfDice(die_priors=[0.3770510623276052, 0.6229489376723948], dice=[Die([0.2981, 0.3018, 0.2004, 0.1997]), Die([0.2001, 0.1975, 0.2017, 0.4007])]))
(33, BagOfDice(die_priors=[0.4162875875022872, 0.5837124124977127], dice=[Die([0.29

Answer: No, as shown above, changing the convergence criterion to 10^-8 didn't change the estimated parameters closer to the ground truth.